src/spaces/compartment.jl

```diff
@@ -0,0 +1,188 @@
+export CompartmentSpace
+
+struct CompartmentSpace{D,F} <: AbstractSpace
+  grid::GridSpace
+  update_vel!::F
+  dims::NTuple{D, Int}
+  periodic::Bool
+  D::Int
+  spacing::Float64
+end
+
+defvel(a, m) = nothing
+
+function CompartmentSpace(d::NTuple{D,Real}, spacing;
+  update_vel! = defvel, periodic = true) where {D}
+  s = GridSpace(floor.(Int, d ./ spacing), periodic=periodic, metric=:euclidean)
+  return CompartmentSpace{D,typeof(update_vel!)}(s, 
+    update_vel!, size(s),periodic, D, spacing)
+end
+
+"""
+    random_position(model) → pos
+Return a random position in the model's space (always with appropriate Type).
+"""
+function random_position(model::ABM{A, <:CompartmentSpace}) where {A}
+  pos = Tuple(rand(model.space.D) .* model.space.dims)
+end
+
+pos2cell(pos::Tuple) = ceil.(Int, pos)
+pos2cell(a::AbstractAgent) = pos2cell(a.pos)
+
+function add_agent_to_space!(a::A, model::ABM{A,<:CompartmentSpace}) where 
+  {A<:AbstractAgent}
+  push!(model.space.grid.s[pos2cell(a)...], a.id)
+  return a
+end
+
+function remove_agent_from_space!(a::A, model::ABM{A,<:CompartmentSpace}) where 
+  {A<:AbstractAgent}
+  prev = model.space.grid.s[pos2cell(a)...]
+  ai = findfirst(i -> i == a.id, prev)
+  deleteat!(prev, ai)
+  return a
+end
+
+function move_agent!(a::A, pos::Tuple, model::ABM{A,<:CompartmentSpace}) where 
+  {A<:AbstractAgent}
+  remove_agent_from_space!(a, model)
+  a.pos = pos
+  add_agent_to_space!(a, model)
+end
+
+"""
+    move_agent!(agent::A, model::ABM{A, CompartmentSpace}, dt = 1.0)
+Propagate the agent forwards one step according to its velocity,
+_after_ updating the agent's velocity (see [`CompartmentSpace`](@ref)).
+Also take care of periodic boundary conditions.
+
+For this continuous space version of `move_agent!`, the "evolution algorithm"
+is a trivial Euler scheme with `dt` the step size, i.e. the agent position is updated
+as `agent.pos += agent.vel * dt`.
+
+Notice that if you want the agent to instantly move to a specified position, do
+`move_agent!(agent, pos, model)`.
+"""
+function move_agent!(agent::A, model::ABM{A, <: CompartmentSpace}, dt::Real = 1.0) where {A <: AbstractAgent}
+  model.space.update_vel!(agent, model)
+  pos = agent.pos .+ dt .* agent.vel
+  if model.space.periodic
+    pos = mod.(pos, model.space.dims)
+  end
```

Comment:
we should move this to `move_agent!(agent, pos, model)` to ensure that periodicity is taken care of at only one, central, place in the source code.